File: app/src/main/java/com/lazycouple/restapiclient/ui/data/CustomResponse.java
package com.lazycouple.restapiclient.ui.data;

import android.os.Parcel;
import android.os.Parcelable;

/**
 * Created by noco on 2016-10-26.
 */
public class CustomResponse implements Parcelable {
    public int code;
    public String body;
    public String errorbody;

    public CustomResponse() {
    }

    protected CustomResponse(Parcel in) {
        code = in.readInt();
        body = in.readString();
        errorbody = in.readString();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(code);
        dest.writeString(body);
        dest.writeString(errorbody);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<CustomResponse> CREATOR = new Creator<CustomResponse>() {
        @Override
        public CustomResponse createFromParcel(Parcel in) {
            return new CustomResponse(in);
        }

        @Override
        public CustomResponse[] newArray(int size) {
            return new CustomResponse[size];
        }
    };


    public static final class CustomResponseBuilder {
        private String errorbody;
        private int code;
        private String body;

        public CustomResponseBuilder() {
        }

        public static CustomResponseBuilder aCustomResponse() {
            return new CustomResponseBuilder();
        }

        public CustomResponseBuilder errorbody(String errorbody) {
            this.errorbody = errorbody;
            return this;
        }

        public CustomResponseBuilder code(int code) {
            this.code = code;
            return this;
        }

        public CustomResponseBuilder body(String body) {
            this.body = body;
            return this;
        }

        public CustomResponse build() {
            CustomResponse customResponse = new CustomResponse();
            customResponse.body = this.body;
            customResponse.code = this.code;
            customResponse.errorbody = this.errorbody;
            return customResponse;
        }
    }
}

File: app/src/main/kotlin/com/lazycouple/restapiclient/ui/data/CustomResponse.kt
package com.lazycouple.restapiclient.ui.data

import android.os.Parcel
import android.os.Parcelable

/**
 * Created by noco on 2016-10-26.
 */
class CustomResponse : Parcelable {
    var code: Int = 0
    var body: String? = null
    var errorbody: String? = null

    constructor() {}

    protected constructor(`in`: Parcel) {
        code = `in`.readInt()
        body = `in`.readString()
        errorbody = `in`.readString()
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(code)
        dest.writeString(body)
        dest.writeString(errorbody)
    }

    override fun describeContents(): Int {
        return 0
    }


    class CustomResponseBuilder {
        private var errorbody: String? = null
        private var code: Int = 0
        private var body: String? = null

        fun errorbody(errorbody: String): CustomResponseBuilder {
            this.errorbody = errorbody
            return this
        }

        fun code(code: Int): CustomResponseBuilder {
            this.code = code
            return this
        }

        fun body(body: String): CustomResponseBuilder {
            this.body = body
            return this
        }

        fun build(): CustomResponse {
            val customResponse = CustomResponse()
            customResponse.body = this.body
            customResponse.code = this.code
            customResponse.errorbody = this.errorbody
            return customResponse
        }

        companion object {

            fun aCustomResponse(): CustomResponseBuilder {
                return CustomResponseBuilder()
            }
        }
    }

    companion object {

        val CREATOR: Parcelable.Creator<CustomResponse> = object : Parcelable.Creator<CustomResponse> {
            override fun createFromParcel(`in`: Parcel): CustomResponse {
                return CustomResponse(`in`)
            }

            override fun newArray(size: Int): Array<CustomResponse?> {
                return arrayOfNulls<CustomResponse>(size)
            }
        }
    }
}
